### PR TITLE
Improve hubert recipe for pre-training and fine-tuning

### PR DESCRIPTION
**Pull request by nateanl**
following pr #2716
- For preprocessing
  - The HuBERT feature takes lots of memory which may not fit some machines. Enable to use a subset of feature for training a k-means model.

- For pre-training
  - Normalize the loss based on the total number of masked frames across all GPUs.
  - Use mixed precision training. fp16 is not well supported in pytorch_lightning.
  - Log accuracies of masked/unmasked frames during training.
  - Clip the gradients with norm `10.0`.

- For ASR fine-tuning
  - Normalize the loss based on the total number of batches across all GPUs, same as in the conformer recipe of TorchAudio.
  - Use mixed precision training.
  - Add "|" after the end of transcription to capture the silence/word termination, same as in fairseq recipe.

- Update the WER results on LibriSpeech dev and test sets.

|                   | WER% (Viterbi)|  WER% (KenLM) |
|:-----------------:|--------------:|--------------:|
| dev-clean         |       10.9    |       4.2     |
| dev-other         |       17.5    |       9.4     |
| test-clean        |       10.9    |       4.4     |
| test-other        |       17.8    |       9.5     |

### CODE REVIEW

File: examples/hubert/evaluate.py [hunk 1]
```diff
@@ -138,8 +144,8 @@ def _parse_args():
     )
     parser.add_argument(
         "--beam-size-token",
-        type=Optional[int],
-        default=None,
+        type=int,
+        default=29,
         help="Number of tokens to consider at each beam search step. (Default: None)",
```
Comment (carolineechen) [inline examples/hubert/evaluate.py:149]:
```suggestion
        help="Number of tokens to consider at each beam search step. (Default: 29)",
```

File: examples/hubert/finetune.py [hunk 2]
```diff
@@ -101,6 +105,18 @@ def _parse_args():
         required=True,
         help="Path to the LibriSpeech and LibriLightLimited datasets.",
     )
+    parser.add_argument(
+        "--checkpoint",
+        type=str,
+        required=True,
+        help="Path to the pre-trained HuBERTPretrainModel checpoint as the initialization.",
```
Comment (carolineechen) [inline examples/hubert/finetune.py:112]:
```suggestion
        help="Path to the pre-trained HuBERTPretrainModel checkpoint as the initialization.",
```

File: examples/hubert/preprocess.py [hunk 3]
```diff
@@ -60,6 +60,12 @@ def _parse_args():
         type=int,
         help="The number of clusters for KMeans clustering.",
     )
+    parser.add_argument(
+        "--percent",
+        default=-1,
+        type=float,
+        help="The number of clusters for KMeans clustering.",
```
Comment (carolineechen) [inline examples/hubert/preprocess.py:67]: should this be something like this:

```suggestion
        help="The percent of data for KMeans clustering. If negative, use all data. (Default: -1)",
```

File: examples/hubert/train.py [hunk 4]
```diff
@@ -90,7 +92,7 @@ def _parse_args():
     )
     parser.add_argument(
         "--resume-checkpoint",
-        type=Optional[pathlib.Path],
+        type=pathlib.Path,
```
Comment (carolineechen) [inline examples/hubert/train.py:95]: This should remain `Optional[pathlib.Path]`

Comment (nateanl) [reply examples/hubert/train.py:95]: actually using `Optional` in argparse will throw an error, that's why I removed those from the parser.

**Review by carolineechen — APPROVED** [review]

**Comment by arlofaria** [comment]
Looking at this now, I have a few thoughts...

> Normalize the loss based on the total number of masked frames across all GPUs.

Could you help me to understand the rationale for this change?  From my understanding, the prior code was doing a reasonable thing: a GPU that processes fewer masked frames should contribute less to the DDP-averaged gradient than a GPU that processes more masked frames.  Moreover, normalizing the loss now means that the predetermined learning rates from the original HuBERT paper might be on the wrong scale; from what I can tell of the paper and the Fairseq code, it doesn't seem like per-masked-frame loss normalization was used, right?

However, for logging purposes, the normalized loss can be very convenient as it allows experiments with different batch sizes to have their curves compared on the same scale.

> Use mixed precision training. fp16 is not well supported in pytorch_lightning.

Perhaps that's a bug that has since been fixed?  I tried today with `lightning==1.8.3` and setting `Trainer(precision=16)` seems to work correctly (same loss curves per step, but at faster convergence per walltime).

> Log accuracies of masked/unmasked frames during training.

Very nice, thanks for the improvement!  However, why are the stats reset only after every epoch (at the end of the validation step), rather than reset for every training step?  This leads to somewhat confusing curves that tend toward the cumulative epoch-level accuracy for steps that are late in an epoch, but are generally under-reporting the more instantaneous step-level accuracy.

> Clip the gradients with norm 10.0.

(nit) This is configurable for pre-training, but not for fine-tuning where it is hardcoded to 5.0.

**Comment by nateanl** [comment]
> it doesn't seem like per-masked-frame loss normalization was used, right?

It actually does in fairsq, not in hubert loss, but in the general fairseq trainer. Here is the code point:
https://github.com/facebookresearch/fairseq/blob/c19aed8ef59acafc8c146b50aed0ff3275ffcf01/fairseq/trainer.py#L932-L946

> Perhaps that's a bug that has since been fixed?

Yes, after fixing the embedding dtype, pytorch_lightning's fp16 API can also be used.

> However, why are the stats reset only after every epoch (at the end of the validation step), rather than reset for every training step?

This is to estimate more reasonable accuracy for each epoch. It indeed introduces cumulative accuracy for each step which has momentum of previous steps. maybe a better way is to report the step accuracy based on the current batch, and use a separate variable to log the total number of count and accuracy sum for the epoch-level accuracies.

> This is configurable for pre-training, but not for fine-tuning where it is hardcoded to 5.0.

Yes, hyperparamters is always hard to tune. In my experiments 10 seems reasonable for pre-training, but 5.0 works better for fine-tuning.

**Comment by arlofaria** [comment]
> It actually does in fairsq, not in hubert loss, but in the general fairseq trainer.

Oh, you're absolutely right -- thanks for looking into this and clarifying the reason for the loss normalization!
